### DISCUSSION TITLE
Migrate to Material "Symbols"

### DISCUSSION THREAD

**Discussion by LasseRosenow** [directus/directus#12958] [Feature Requests]
There is a new Material Icons collection called "Material Symbols"

https://fonts.google.com/icons?icon.set=Material+Symbols

As far as I understand it, there are a few more icons, some look different and also they are much more customizable than the old icons.

Also it seems to finally fix all those weird inconsistencies like having a filled "help" icon inside the "outlined" font :D

Is there any plan to migrate?

**Comment by rijkvanzanten** [comment]
Do you happen to know if the icons are "backwards compatible"? (Eg use the same names as the previous material icons) 🤔

**Reply by benhaynes** [reply]
Ohh, it's a variable font too... so much more flexibility!

**Reply by rijkvanzanten** [reply]
Fun to expose those icon variables through theming too (cc @AustinPhillipTaylor ✨)

**Reply by benhaynes** [reply]
Also, it'd be nice to have a formal (no-code) way to extend the base icon set with custom SVG options (similar to Discord). This could also exist within the upcoming theming panel.

**Reply by AustinPhillipTaylor** [reply]
I hadn't heard about Material Symbols, but it looks fantastic! It'll solve a lot of the problems I mentioned before, regarding inconsistent icon weights. 

This is definitely something we should explore.

**Reply by AustinPhillipTaylor** [reply]
Also @rijkvanzanten, to answer your question, I took a look at some of the icon names and it appears that there is a minor discrepancy in the naming. For instance, in our Module bar we use the `help_outline` icon for the docs link. In Material Symbols this icon is just named `help`. This makes sense, since it eliminates the need to have explicit variations for each icon type. Otherwise, as far as I can tell it looks like the base names all line up.

**Reply by rijkvanzanten** [reply]
Doing a find and replace on `_outline` isn't the end of the world, but we have to be careful as people could have configured those icons for other things (like collections/bookmarks)

**Reply by AustinPhillipTaylor** [reply]
Oh, true. Fortunately, we have all icons incorporated through the `v-icon` component, so I don't think it'd be too difficult to run a quick check `iconName.endsWith('_outline') && iconName.slice(0, -8)`

**Reply by rijkvanzanten** [reply]
Nah that's a little too smelly for my taste. I'd prefer doing a single migration to make sure everything is up to date and correct, and prevent any semi-backwards-compatible hacks in the codebase itself

**Reply by LasseRosenow** [reply]
There needs to be some sort of migration. It is not only limited to outline.

For example `123` does not exist or is named differently in the new one and there are probably more

**Reply by AustinPhillipTaylor** [reply]
Yeah, that's a strange one. It looks like they just removed that icon. Closest I could find is `Pin`. Regardless, if they removed the icon, that's not something that could be fixed with a migration.

**Reply by rijkvanzanten** [reply]
nooooooo that makes it way more annoying to get this upgraded :C Is there a list available somewhere of all the known changes?

**Reply by AustinPhillipTaylor** [reply]
I see your point, regarding not wanting to add a hacky check in the component, but I feel like trying to migrate all of the icon names is a little dangerous. They would have to be changed in the icon fields of course, but also in all of the options of interfaces that implement icons (i.e. `iconLeft` and `iconRight` on inputs, etc.). When you start getting into nested JSON, I personally feel like it's just a problem waiting to happen. 

That also wouldn't account for icon usage in custom modules, it would force devs to update those still. 

One possible solution that would solve all of the issues relatively easily, simply implement the Material Symbols, then set the Material Icons as a fallback font. Not hacky, removed icons can be persisted, and it doesn't require any migration.

**Reply by rijkvanzanten** [reply]
I think that has similar long term maintenance issues 🤔  I'd rather have a clean one-time upgrade and make sure there's no tech debt after that 👍🏻

**Reply by AustinPhillipTaylor** [reply]
My concern is attempting to modify people's data. If it was as simple as modifying a single value in a specific string column, it wouldn't bother me. However, icon usage is a bit unpredictable, especially on repeaters or custom interfaces. Maybe I'm just overly cautious.

Regardless, assuming we do go the migration route, it still doesn't solve the issue of missing/removed icons.

**Reply by AustinPhillipTaylor** [reply]
Actually, we can scrap this entire conversation. I just tested it, Material Symbols is actually entirely backwards compatible. All variants automatically fall back to the base icon. We don't need to do anything but change the font name 🤯 
And import the font, of course.

**Reply by AustinPhillipTaylor** [reply]
It even looks like `123` is still part of the symbol set, just doesn't show up on their icons page https://github.com/google/material-design-icons/blame/master/update/current_versions.json#L2148

**Reply by rijkvanzanten** [reply]
> Actually, we can scrap this entire conversation. I just tested it, Material Symbols is actually entirely backwards compatible. All variants automatically fall back to the base icon. We don't need to do anything but change the font name 🤯
> And import the font, of course.

That's what we like to hear 🎉

**Reply by AustinPhillipTaylor** [reply]
We'll still probably want to find and replace all `_outline` throughout the app, just for the sake of consistency, but it's nice that it won't be a necessity for this to work!

**Reply by hanneskuettner** [reply]
After playing around with it and actually attempting to replace it I made the following observations:

1. Material Symbols adds a total of 991 new icons 🎉 
2. The variable font is a massive *3MB*, I was able to downsize it by only including the previously Axis values (wght 400, opsz 24) to ~300kb.
3. There are indeed fallback codepoints in the font that are either spiritual successors, eg. [`add_ic_call`](https://fonts.google.com/icons?selected=Material%20Icons%20Outlined%3Aadd_ic_call%3A) -> [`add_call`](https://fonts.google.com/icons?selected=Material%20Symbols%20Outlined%3Aadd_call%3AFILL%400%3Bwght%40400%3BGRAD%400%3Bopsz%4048), or replacing the duplicate `outline` and `filled` variants with a single variant. 
4. There are a total of 21 real removals out of a total of 2007 icons that don't have a fallback defined in the font (listed below). Most of are either super unique, brand logos or seem to be a bug in their codepoints, since eg. `panorama_horizontal_select` at least to me have a clear replacement.

<details><summary>List of removals</summary>

```
catching_pokemon
delivery_dining
discount
emoji_events
fitbit
highlight_alt
miscellaneous_services
moped
motion_photos_on
no_cell
panorama_horizontal_select
panorama_photosphere_select
panorama_vertical_select
panorama_wide_angle_select
person_add_alt_1
person_remove_alt_1
pix
play_circle_filled
play_circle_outline
signal_wifi_statusbar_connected_no_internet_4
whatsapp
```

</details>

The second point really isn't a problem if we want to have a status quo (without using any actual variability for now), but having a 300kb font file.

The third point has some slight implications, since for example the icon `bookmark` in Material Symbols is an outline by default now instead of filled. This will lead to some visual changes in the icon when the outlined fallback is used. But since all those icons are the same semantically mostly identical I don't think that is too much of a problem?

The last point is the actual headache, but I would love to know many actually use these 21 icons. Any idea how to proceed with that?

In the meantime I've opened a PR #18004

**Reply by paescuj** [reply]
Thank you very much for your research and contribution! ❤️ 

> The second point really isn't a problem if we want to have a status quo (without using any actual variability for now), but having a 300kb font file.

I'd very much prefer the 300kb file 😄 

> The third point has some slight implications, since for example the icon bookmark in Material Symbols is an outline by default now instead of filled. This will lead to some visual changes in the icon when the outlined fallback is used. But since all those icons are the same semantically mostly identical I don't think that is too much of a problem?

First and foremost, we need to make sure that the app is not affected by such visual changes. Do we know which icons (variants) are affected by this?

> The last point is the actual headache, but I would love to know many actually use these 21 icons. Any idea how to proceed with that?

I wouldn't really consider that a problem. Firstly, those seem to be mostly some exotic icons and there are reasons why those got removed and secondly, we can point out these changes in the release notes to give users a chance to identify & replace them.

**Reply by rijkvanzanten** [reply]
> I wouldn't really consider that a problem. Firstly, those seem to be mostly some exotic icons and there are reasons why those got removed and secondly, we can point out these changes in the release notes to give users a chance to identify & replace them.

Same here 👍🏻  Worst case scenario, the end users loses an icon to use in the UI. They're just used for visual presentation, and don't really affect the usage of the system in any other meaningful way.

**Reply by hanneskuettner** [reply]
> I'd very much prefer the 300kb file 😄

Agreed. Having played a bit more with it, there is also the option to [keep the `opsz` axis](https://developers.google.com/fonts/docs/material_symbols#opsz_axis) fully available for icon sizes from `20px` to `48p` and arrive at ~600kb. 

> First and foremost, we need to make sure that the app is not affected by such visual changes. Do we know which icons (variants) are affected by this?

I've compiled a list below of all the icons that are either completely deprecated (while working through this I noticed that the app itself uses two icons `copy` and `bookmark_outline` which aren't officially documented in Material *Icons* but still have a fallback in the icon font) or have been renamed.
The completely deprecated ones are only a problem in the app source itself since they have never been allowed as valid selections in the icon-select.

I've compiled a list of all used icons (and probably a few false positives, since I only searched for `'[iconName']` and `"[iconName]"` in the code base).

[Comparison](https://github.com/directus/directus/files/11116084/compare.pdf)

As you can see, as expected are almost all of them visually identical (or very close).
The red ones need action since they aren't available in the new font anymore (3 icons in total)

`bookmark_outline` -> `bookmark`
`copy` -> `content_copy`
`highlight_alt` -> probably `select`?

The other ones that we need to pay attention to are the icons that are filled in the left column since they were filled by default in the old icon set. They won't be filled anymore now and should receive the `filled` prop and then they're fine again.

As far as I can see those are `bookmark`, `error`, `star` and `warning`.

I will include those changes in my PR asap.

Additionally all icons marked in yellow have been removed from the official metadata and are only fallbacks in the font. We should update those to the canonical names once we're at it.

If you want I can make that part of the PR as well.

> I wouldn't really consider that a problem. Firstly, those seem to be mostly some exotic icons and there are reasons why those got removed and secondly, we can point out these changes in the release notes to give users a chance to identify & replace them.

Perfect. That's what I would have suggested as well.

**Reply by LasseRosenow** [reply]
Do the Icons that used to be filled really need attention?
The way I see it is, that them being filled in the old icon set is a huge inconsistency problem since all else are outlined hence the change.

So them now just being outlined also would just increase consistency from my point of view

**Reply by rijkvanzanten** [reply]
> Do the Icons that used to be filled really need attention?
> 
> The way I see it is, that them being filled in the old icon set is a huge inconsistency problem since all else are outlined hence the change.
> 
> 
> 
> So them now just being outlined also would just increase consistency from my point of view

I agree. Only thing I'd like to check is that we didn't use the outline vs filled as an indicator for something, but I don't think we did

**Reply by hanneskuettner** [reply]
> I agree. Only thing I'd like to check is that we didn't use the outline vs filled as an indicator for something, but I don't think we did

Looks like the filled bookmark is used as an indicator

https://github.com/directus/directus/blob/953c2f95d57a944ccb4771e41505b8d5caa80189/app/src/modules/content/routes/collection.vue#L54

And the star should be filled for the rating display.

The others seem to be fine?

**Reply by hanneskuettner** [reply]
One potential problem I discovered while working on this a bit more.

The directus_presents icon field uses the `bookmark_outline` icon, that was already deprecated and not officially documented in the Material Icons font but had a fallback defined to fall back to `bookmark_border`. So it worked as intended (TM) if the icon value was used. With Material Symbols this fallback is dropped.
Now there are probably quite a few presets with the `bookmark_outline` icon in the presets table. 

Since that that is *our* table are we comfortable with running a migration to replace all instances of `bookmark_outline` in with `bookmark`?

**Reply by hanneskuettner** [reply]
Another caveat I noticed.

The old, Material Icons has no glyphs for the alphabet, the new one does. 
This leads to problems where in the app (not in user content) icons were used that never officially existed, eg. `info_outline`, `arrow_backwards` but worked because they were prefixes of valid icon names, so only the ligature making up the existing icon was shown and the suffix was silently dropped, eg. `info_outline` -> `info`, `arrow_backwards` -> `arrow_back`.

In the new font the glyphs for the alphabet are defined, leading to ugly, but very noticeable mistakes. 

![Screenshot 2023-04-02 at 11 36 51](https://user-images.githubusercontent.com/4376726/229344849-fc1ca453-a4ac-4e1d-92b9-1b0c15f223fe.png)

My approach now is that I will update my code search to find all those instances as well and get rid of them for good.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
